### PR TITLE
Assignment6-Joel

### PR DESCRIPTION
**Pull request by RSXEnthusiast**
Finished... I hope.

### CODE REVIEW

File: Assignment6/MyArray.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Assignment6 {
+    public class MyArray<Data> : ICollection<Data> {
```
Comment (mmwoodfo) [inline Assignment6/MyArray.cs]:
```suggestion
    public class MyArray<TCollection> : ICollection<TCollection> {
```
You should be getting a warning for this as the generic type is usually prefixed with a T. Calling your generic Data may be confusing.

File: Assignment6/MyArray.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        private Data[] _Storage;
+        public int Capacity { get; set; }
```
Comment (mmwoodfo) [inline Assignment6/MyArray.cs]:
```suggestion
        public int Capacity { get; }
```
You should be able to make this read-only if you want.

File: Assignment6/MyArray.cs (hunk 1 above)
Comment (RSXEnthusiast) [reply Assignment6/MyArray.cs]: No error, but I'll fix that. Thanks!

File: Assignment6/MyArray.cs (hunk 2 above)
Comment (RSXEnthusiast) [reply Assignment6/MyArray.cs]: Makes sense to me. Thanks!

File: Assignment6/MyArray.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace Assignment6 {
+    public class MyArray<TData> : ICollection<TData> {
+        private TData[] _Storage;
+        public int Capacity { get; }
+        public int Count { get; set; }
+
+        public MyArray(int capacity) {
+            if (capacity <= 0) {
+                throw new ArgumentOutOfRangeException("capacity");
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: Use nameof in place of string literal 'capacity'

File: Assignment6/MyArray.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            Count = 0;
+            Capacity = capacity;
+            _Storage = new TData[capacity];
+        }
+
+        public bool IsReadOnly => _Storage.IsReadOnly;
+
+        public void Add(TData item) {
+            if (item is null) {
+                throw new ArgumentNullException("item");
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: Use nameof in place of string literal 'item'

File: Assignment6/MyArray.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        public void Clear() {
+            _Storage = new TData[Capacity];
+            Count = 0;
+        }
+
+        public bool Contains(TData item) {
+            if (item is null) {
+                throw new ArgumentNullException("item");
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: Use nameof in place of string literal 'item'

File: Assignment6/MyArray.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                throw new NullReferenceException("Array passed into CopyTo is null!");
+            }
+            for (int i = arrayIndex, j = 0; i < _Storage.Length; i++, j++) {
+                array[j] = _Storage[i];
+            }
+        }
+
+        public bool Remove(TData item) {
+            if (item is null) {
+                throw new ArgumentNullException("item");
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: Use nameof in place of string literal 'item'

File: Assignment6/MyArray.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                throw new ArgumentOutOfRangeException("capacity");
+            }
+            Count = 0;
+            Capacity = capacity;
+            _Storage = new TData[capacity];
+        }
+
+        public bool IsReadOnly => _Storage.IsReadOnly;
+
+        public void Add(TData item) {
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: This functionality is not typical to an array.  Consider implementing ICollection<T>'s Add/Remove methods explicitly instead.

File: Assignment6/MyArray.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            }
+            if (array is null) {
+                throw new NullReferenceException("Array passed into CopyTo is null!");
+            }
+            for (int i = arrayIndex, j = 0; i < _Storage.Length; i++, j++) {
+                array[j] = _Storage[i];
+            }
+        }
+
+        public bool Remove(TData item) {
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: This functionality is not typical to an array.  Consider implementing ICollection<T>'s Add/Remove methods explicitly instead.

File: Assignment6/MyArray.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Assignment6 {
+    public class MyArray<TData> : ICollection<TData> {
+        private TData[] _Storage;
+        public int Capacity { get; }
+        public int Count { get; set; }
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: Why the public setter?  This would seem to break encapsulation?

File: Assignment6/MyArray.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Assignment6 {
+    public class MyArray<TData> : ICollection<TData> {
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: I'm not seeing an index operator implementation.

File: Assignment6/MyArray.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                array[j] = _Storage[i];
+            }
+        }
+
+        public bool Remove(TData item) {
+            if (item is null) {
+                throw new ArgumentNullException("item");
+            }
+            for (int i = 0; i < _Storage.Length; i++) {
+                //There's an error saying _Storage[i] could be null, but so what if it is?
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: You want to use the null forgiveness operator for this.

File: Assignment6/MyArray.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                    //It says data is non-nullable, but it's definetly able to set _Storage[i] to null...
+                    _Storage[i] = default;
+                    Count--;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public TData GetData(int index) {
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: This is where you want to have an index operator instead.

File: Assignment6/MyArray.cs [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            if (index < 0 || index >= Capacity) {
+                throw new IndexOutOfRangeException("Index in GetData is out of bounds!");
+            }
+            if (_Storage[index] is null) {
+                throw new NullReferenceException("_Storage[index] is null!");
+            }
+            return _Storage[index];
+        }
+
+        public IEnumerator<TData> GetEnumerator() {
```
Comment (MarkMichaelis) [inline Assignment6/MyArray.cs]: Why don't you just use `IEnumerator IEnumerable.GetEnumerator() => _Storage.GetEnumerator();`?

File: Assignmnet6.tests/ArrayTests.cs [hunk 14]
```diff
@@ -0,0 +1,309 @@
+using Assignment6;
```
Comment (MarkMichaelis) [inline Assignmnet6.tests/ArrayTests.cs]: You file name should match the SUT class you are testing but with a "Tests" suffix.

File: Assignmnet6.tests/ArrayTests.cs [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+        [TestMethod]
+        public void IEnumerator_Works() {
+            //Arrange
+            int[] a = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+            MyArray<int?> array = CreateIntMyArray(10, a);
+            ArrayList list = new ArrayList();
+
+            //Act
+
+            foreach (int? item in array) {
```
Comment (MarkMichaelis) [inline Assignmnet6.tests/ArrayTests.cs]: I'm pleased to see you test the iterator with a foreach.

**Comment by MarkMichaelis** [comment]
##  (RSXEnthusiast)

## Assignment6

| Build/Compile (Build Succeeded: 0 Error(s), **5 Warning(s)**)| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5** |


| Assignment  | **12** |
| --------------------- | ------ |
| Supports collecting a fixed number of items once instantiated. | ✔ |
| Has a constructor via which you can specify fixed width.  (Other constructors are welcome.) | ✔ |
| Handle null appropriately when items are missing (this is likely a judgement call so justify your answer in the comments.) | ✔ |
| Throw an exception if an item doesn't exist. | ✔ |
| Include an index operator. | ❌ |
| Support using foreach over the items in the `Array<T>` class. | ✔ |
| Support a collection initializer. | ✔ |
| Implement `System.Collections.Generic.ICollection<T>`. | ✔ |
| Ignore support for multi-dimensional arrays. | ✔ |
| Add a `Capacity` property to return the capacity (the value specified in the constructor) of the collection. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **❌** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment6 | Enable | Assignment6.sln |  | CS1701, CS1702, CS2008 | 4 |
| Assignmnet6.Tests | Enable | Assignment6.sln |  | CS1701, CS1702 | 4 |


| Testing | **15** |
| --------------------- | ------ |
| Test Run FAILED: Total tests=22, Passed=21, Failed=1, Skipped=0 | **89%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **0** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write a cast operator to a C# array or a `System.Collections.Generic.List<T>`. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance | ❌ |

**Comment by MarkMichaelis** [comment]
##  (RSXEnthusiast)

## Assignment6

| Build/Compile (Build Succeeded: 0 Error(s), **5 Warning(s)**)| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5** |


| Assignment  | **12** |
| --------------------- | ------ |
| Supports collecting a fixed number of items once instantiated. | ✔ |
| Has a constructor via which you can specify fixed width.  (Other constructors are welcome.) | ✔ |
| Handle null appropriately when items are missing (this is likely a judgement call so justify your answer in the comments.) | ✔ |
| Throw an exception if an item doesn't exist. | ✔ |
| Include an index operator. | ❌ |
| Support using foreach over the items in the `Array<T>` class. | ✔ |
| Support a collection initializer. | ✔ |
| Implement `System.Collections.Generic.ICollection<T>`. | ✔ |
| Ignore support for multi-dimensional arrays. | ✔ |
| Add a `Capacity` property to return the capacity (the value specified in the constructor) of the collection. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **❌** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment6 | Enable | Assignment6.sln |  | CS1701, CS1702, CS2008 | 4 |
| Assignmnet6.Tests | Enable | Assignment6.sln |  | CS1701, CS1702 | 4 |


| Testing | **15** |
| --------------------- | ------ |
| Test Run FAILED: Total tests=22, Passed=21, Failed=1, Skipped=0 | **89%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **0** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write a cast operator to a C# array or a `System.Collections.Generic.List<T>`. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance | ❌ |

**Pull request closed without merging**